src/raftstore/store/store.rs

```diff
@@ -1470,6 +1633,255 @@ impl<T: Transport, C: PdClient> Store<T, C> {
+        self.propose_raft_command(req, Callback::None);
+        Ok(())
+    }
+
+    fn rollback_merge(&mut self, region: &metapb::Region) {
+        let req = {
+            let peer = &self.region_peers[&region.get_id()];
+            let state = peer.pending_merge.as_ref().unwrap();
+            let mut request = new_admin_request(region.get_id(), peer.peer.clone());
+            request
+                .mut_header()
+                .set_region_epoch(peer.region().get_region_epoch().clone());
+            let mut admin = AdminRequest::new();
+            admin.set_cmd_type(AdminCmdType::RollbackMerge);
+            admin.mut_rollback_merge().set_commit(state.get_commit());
+            request.set_admin_request(admin);
+            request
+        };
+        self.propose_raft_command(req, Callback::None);
+    }
+
+    fn on_check_merge(&mut self, event_loop: &mut EventLoop<Self>) {
+        let merging_regions = self.merging_regions.take().unwrap();
+        for region in &merging_regions {
+            if let Err(e) = self.schedule_merge(region) {
+                info!(
+                    "[region {}] failed to schedule merge, rollback: {:?}",
+                    region.get_id(),
+                    e
+                );
+                self.rollback_merge(region);
+            }
+        }
+        self.merging_regions = Some(merging_regions);
+        self.register_merge_check_tick(event_loop);
+    }
+
+    fn on_ready_prepare_merge(&mut self, region: metapb::Region, state: MergeState) {
+        {
+            let peer = self.region_peers.get_mut(&region.get_id()).unwrap();
+            peer.pending_merge = Some(state);
+            peer.mut_store().region = region.clone();
+        }
+
+        if let Err(e) = self.schedule_merge(&region) {
+            info!(
+                "[region {}] failed to schedule merge, rollback: {:?}",
+                region.get_id(),
+                e
+            );
+            self.rollback_merge(&region);
+        }
+        self.merging_regions.as_mut().unwrap().push(region);
+    }
+
+    fn on_ready_commit_merge(&mut self, region: metapb::Region, source: metapb::Region) {
+        let source_peer = {
+            let peer = self.region_peers.get_mut(&source.get_id()).unwrap();
+            // In some case, PrepareMerge entry may be carried by the dst region. So
+            // the peer is not in merging mode.
+            if peer.pending_merge.is_none() {
+                let mut state = MergeState::new();
+                state.set_target(region.clone());
+                peer.pending_merge = Some(state);
+            }
+            peer.peer.clone()
+        };
+        self.destroy_peer(source.get_id(), source_peer, true);
+        // If merge backward, then stale meta is clear when source region is destroyed.
+        // So only forward needs to be considered.
+        if region.get_end_key() == source.get_end_key() {
```

Comment:
how about end key not equal? no need to remove the source from the ranges?

Reply:
`destroy_peer` always remove source from ranges as comment at L1824 says.